sycl/doc/SYCL_compiler_and_runtime_design.md

```diff
@@ -42,6 +42,91 @@ back-end. Today middle-end transformations include just a couple of passes:
 - **Back-end** - produces native "device" code in ahead-of-time compilation
 mode.
 
+### SYCL support in clang frontend
+
+SYCL support in clang frontend can be split into the following components:
+
+- Device code outlining. Since SYCL is a single source programming model
+compiler should be able to separate device code from host code.
+
+- Lowering of lambda function objects and named function objects ("SYCL kernel
+functions"). To execute "SYCL kernel functions" on OpenCL devices some
+transformations are required.
+
+- Device code diagnostics.
+
+- Integration header generation.
+
+#### Device code outlining
+
+Here is a code example of a SYCL program that demonstrates compiler outlining
+work:
+
+```C++
+int foo(int x) { return ++x; }
+int bar(int x) { throw std::exception("CPU code only!"); }
+...
+using namespace cl::sycl;
+queue Q;
+buffer<int, 1> a(range<1>{1024});
+Q.submit([&](handler& cgh) {
+      auto A = a.get_access<access::mode::write>(cgh);
+      cgh.parallel_for<init_a>(range<1>{1024}, [=](id<1> index) {
+        A[index] = index[0] * 2 + index[1] + foo(42);
+      });
+    }
+...
+```
+
+SYCL compiler needs to compile lambda exression passed to
+`cl::sycl::handler::parallel_for` method and function `foo` called from this
+lambda function. Compiler also must ignore bar function when we compile the "device" part
+of the single source code.
+
+Current approach is to use the SYCL kernel atttribute in SYCL runtime to mark code
+passed to `cl::sycl::handler::parallel_for` as "kernel functions".
+Obviously runtime library can't mark foo as "device" code - this is a compiler
+job: to traverse all symbols accessible from kernel functions and add them to
+the "device part" of the code marking them with the new SYCL device attribute.
+
+#### Lowering of lambda function objects and named function objects
+
+All SYCL memory objects shared between host and device (buffers/images,
+these objects map to OpenCL buffers and images) must be accessed through special
+`accessor` classes. The "device" side implementation of these classes contain pointers to the device memory. There is no
+way in OpenCL to pass structures with pointers inside as kernel arguments.
+SYCL also has special mechanism for passing kernel arguments from host to
+device, if in OpenCL you need to call `clSetKernelArg`, in SYCL all
+kernel arguments are captures/fields of lambda/functor which is passed to
+`parallel_for` (in code snippet above one kernel argument - `accessor A`).
+To map to OpenCL kernel arguments setting mechanism we added generation of
+"kernel wrapper" function inside the compiler. "Kernel wrapper" function
+contains body of SYCL kernel function, receives OpenCL like parameters and
+additionally does some manipulation to initialize captured lambda/functor
+fields with these parameters. In some pseudo code "kernel wrapper" looks like
+this:
+
+```C++
+// SYCL kernel is defined in SYCL headers
+__attribute__((sycl_kernel)) someSYCLKernel(lambda) {
```

Comment:
What is the meaning of this "lambda"? Where is the type?
Do you mean something like this:
```
template <typename Lambda>
 __attribute__((sycl_kernel)) someSYCLKernel(Lambda lambda) {
```
?

Reply:
Yes, I added type.